src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,244 @@
+        // We want to compare the current method signature to any others with the exact same arguments in the exact same order.
+        private static bool VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(
+            INamedTypeSymbol cancellationTokenType,
+            IMethodSymbol methodDeclaration,
+            out string? cancellationTokenParameterName)
+        {
+            cancellationTokenParameterName = null;
+
+            if (methodDeclaration.Parameters.Count(x => x.Type.Equals(cancellationTokenType)) == 1 &&
+                methodDeclaration.Parameters.Last() is IParameterSymbol lastParameter &&
+                lastParameter.Type.Equals(cancellationTokenType))
+            {
+                cancellationTokenParameterName = lastParameter.Name;
+            }
+
+            return cancellationTokenParameterName != null;
+        }
+
+        // Check if the currently used overload is the one that takes the ct, but is utilizing the default value offered in the method signature.
+        // We want to offer a diagnostic for this case, so the user explicitly passes the ancestor's ct.
+        private static bool InvocationIgnoresOptionalCancellationToken(INamedTypeSymbol cancellationTokenType, IInvocationOperation invocation, IMethodSymbol method)
+        {
+            if (method.Parameters.Any() && method.Parameters.Last() is IParameterSymbol lastParameter && lastParameter.Type.Equals(cancellationTokenType))
+            {
+                // If the ct parameter has a default value, return true if a value is not being explicitly passed in the invocation
+                return lastParameter.IsOptional && // Has a default value
+                       invocation.Arguments.Last() is IArgumentOperation lastArgument &&
+                       lastArgument.IsImplicit; // The default value is being used
+            }
+
+            return false;
+        }
+
+        // Check if there's a method overload with the same parameters as this one, in the same order, plus a ct at the end.
+        private static bool MethodHasCancellationTokenOverload(ITypeSymbol cancellationTokenType, IMethodSymbol method)
+        {
+            IMethodSymbol? overload = method.ContainingType.GetMembers(method.Name).OfType<IMethodSymbol>().FirstOrDefault(x => HasSameParametersPlusCancellationToken(cancellationTokenType, method, x));
+
+            return overload != null;
+        }
+
+        // Checks if the parameters of the two passed methods only differ in a ct.
+        private static bool HasSameParametersPlusCancellationToken(ITypeSymbol cancellationTokenType, IMethodSymbol originalMethod, IMethodSymbol methodToCompare)
+        {
+            if (!originalMethod.Name.Equals(methodToCompare.Name, StringComparison.Ordinal) ||
+                originalMethod.Equals(methodToCompare) ||
+                methodToCompare.Parameters.Length == 0 ||
+                methodToCompare.Parameters.Length != originalMethod.Parameters.Length + 1 ||
```

Comment:
Compare to 0 seems redundant given the next check
```suggestion
                methodToCompare.Parameters.Length != originalMethod.Parameters.Length + 1 ||
```

Reply:
Updated.